pkg/syncer/storagepool/diskDecommissionController.go

```diff
@@ -0,0 +1,283 @@
+			if err != nil {
+				log.Errorf("Failed to get PV resource %v. Error: %v", pvName, err)
+				migrationFailed = true
+				break
+			}
+			pvcName, _, _ := unstructured.NestedString(pv.Object, "spec", "claimRef", "name")
+			namespace, found, err := unstructured.NestedString(pv.Object, "spec", "claimRef", "namespace")
+			if pvcName == "" || !found || err != nil {
+				log.Errorf("Failed to get PVC bounded to PV %v. Error: %v", pvName, err)
+				migrationFailed = true
+				break
+			}
+			pvc, err := addTargetSPAnnotationOnPVC(ctx, pvcName, namespace, targetSPName)
+			if err != nil {
+				log.Errorf("Failed to add target SP annotation to PVC %v. Error: %v", pvcName, err)
+				migrationFailed = true
+				break
+			}
+			pvcToMigrate = append(pvcToMigrate, pvc)
+		}
+
+		_, unsuccessfulMigrations := w.MigrateVolumes(ctx, pvcToMigrate, true)
+		if len(unsuccessfulMigrations) != 0 {
+			migrationFailed = true
+		}
+	}
+}
+
+func initDiskDecommController(ctx context.Context, migrationCntlr *migrationController) (*DiskDecommController, error) {
+	log := logger.GetLogger(ctx)
+	log.Infof("Starting disk decommission controller")
+	k8sDynamicClient, spResource, err := getSPClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	w := &DiskDecommController{}
+	w.k8sDynamicClient = k8sDynamicClient
+	w.migrationCntlr = migrationCntlr
+	w.spResource = spResource
+	w.pvResource = &schema.GroupVersionResource{Group: "", Version: "v1", Resource: "persistentvolumes"}
+	w.pvcResource = &schema.GroupVersionResource{Group: "", Version: "v1", Resource: "persistentvolumeclaims"}
+	w.diskDecommMode = make(map[string]string)
+	w.execSemaphore = semaphore.NewWeighted(1)
+
+	// get all the pvc resource for which targetSPAnnotationKey annotations exists.
+	// This will give us list of all pending migrations.
+	pvcList, err := w.k8sDynamicClient.Resource(*w.pvcResource).Namespace(v1.NamespaceAll).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return w, err
+	}
+
+	pvcToMigrate := make([]*unstructured.Unstructured, 0)
+	for _, pvc := range pvcList.Items {
+		targetSPName, _, _ := unstructured.NestedString(pvc.Object, "metadata", "annotations", targetSPAnnotationKey)
+		if targetSPName != "" {
+			pvcToMigrate = append(pvcToMigrate, &pvc)
+		}
+	}
+	w.MigrateVolumes(ctx, pvcToMigrate, false)
+
+	// start StoragePool watch to look for events putting SP under disk decommission
+	err = w.renewStoragePoolWatch(ctx)
+	if err != nil {
+		return w, err
+	}
+	go w.watchStoragePool(ctx)
+
+	// get all sp resource
+	spList, err := w.k8sDynamicClient.Resource(*w.spResource).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return w, err
+	}
+	// make progress on StoragePool which are under disk decommission
+	for _, sp := range spList.Items {
+		spName := sp.GetName()
+		if w.shouldEnterDiskDecommission(ctx, &sp) && spName != "" {
+			maintenanceMode := w.diskDecommMode[spName]
+			go w.DecommissionDisk(ctx, spName, maintenanceMode)
+		}
+	}
```

Comment:
In case of syncer restart, the watch may not receive an event for SP under disk decomm. Thus to resume disk decommission after restrt we are iterating through SP to find if any SP was in disk decomm.